### PR TITLE
OAuth: Support `oauth2` as an authentication method

### PR DESCRIPTION
**Pull request by mbana**
Parse and configure OAuth2 option.

`internal/envoy/auth`
---------------------

Introduce new package with some code that used to live in `internal/controllers/parser.go`.

`internal/parser`
-----------------

Restructure code so it calls into `internal/envoy/auth` for auth configuration.

`smoketests/auth_oauth2`
-----------------------

Add smoketests.

`docs/docs/guides/oauth2.md`
----------------------------

Add OAuth2 guide. And a huge caution indicating that the feature is under active development.

Screenshot from 2022-08-19 13-41-19

`docs/reference/extension.md`
-----------------------------

* Update to include OAuth 2.0 configuration.
* Update documentation to include details on how to enable oauth.

`examples/**`
-----------------------

Restructure directory and add `examples/auth/oauth2/authorization-code-grant` example.

`Makefile`
----------

Remove `docker-images-cache` and `docker-build-manager-debug`

References
----------

* <https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_filters/oauth2_filter>

Notes
-----

> The OAuth2 filter/feature is currently under active development.

Signed-off-by: Mohamed Bana <mohamed@bana.io>

---

This PR...

## Changes

-

## Fixes

-

## Checklist

- [ ] tested locally
- [ ] added new dependencies
- [ ] updated the docs
- [ ] added a test

### CODE REVIEW

**Comment by mbana** [comment]
OAuth2 is blocked on the folks at Envoy fixing an upstream issue with Envoy Proxy itself. Please see:

* <https://github.com/envoyproxy/envoy/issues/22678>
* <https://github.com/envoyproxy/go-control-plane/issues/581>

We can't progress further until the above have been fixed.

File: Makefile
Comment (jasmingacic) [inline Makefile:180]: thanks for cleaning up unused sections

File: docs/docs/guides/oauth2.md
Comment (jasmingacic) [inline docs/docs/guides/oauth2.md:5]: Can you add reference to the problem we've seen. Envoy GH issue etc.
And describe the problem we have

File: examples/auth/oauth2/keycloak/keycloak.sh
Comment (jasmingacic) [inline examples/auth/oauth2/keycloak/keycloak.sh:2]: are we using keycloak here?

**Review by jasmingacic — DISMISSED** [review]
Looking good. 

I left few minor  comments but everything else looks sane.

File: examples/auth/oauth2/keycloak/keycloak.sh
Comment (mbana) [reply examples/auth/oauth2/keycloak/keycloak.sh:2]: I'll remove the stale reference. Sorry about that. I was going to experiment with Keycloak at somepoint.

File: docs/docs/guides/oauth2.md
Comment (aabedraba) [inline docs/docs/guides/oauth2.md:66]: Can we not use Kubernetes manifests for examples? 

I think it's better if we stay consistent by just putting the openapi definition as example, which is what we are rooting for. Then we can show them how to apply it with 

`kusk api generate -i openapi.yaml | kubectl apply -f -`

File: docs/docs/reference/extension.md
Comment (aabedraba) [inline docs/docs/reference/extension.md:346]: TODO here

File: docs/docs/guides/oauth2.md
Comment (mbana) [reply docs/docs/guides/oauth2.md:5]: I've done this in a section called `Upstream Issues` now.

File: docs/docs/reference/extension.md
Comment (mbana) [reply docs/docs/reference/extension.md:346]: Noted. Thanks.

File: docs/docs/guides/oauth2.md
Comment (jasmingacic) [inline docs/docs/guides/oauth2.md:35]: is this intentionally left blank?

**Review by jasmingacic — APPROVED** [review]
Let's package this

File: docs/docs/guides/oauth2.md
Comment (mbana) [reply docs/docs/guides/oauth2.md:35]: I am currently writing it, so yes.

File: docs/docs/reference/extension.md
Comment (mbana) [reply docs/docs/reference/extension.md]: How about now?

**Comment by mbana** [comment]
@aabedraba, @jasmingacic and @kylehodgetts , the documentation is now ready.

You can preview it by running `make docs-preview` at the root of the repository. It will launch a browser with the docs, please be patient because it takes a while for `npm install` to complete.

**Review by jasmingacic — APPROVED** [review]
Good to go!!!

Please resolve the conflicts and get this merged.

File: cmd/manager/main.go
Comment (kylehodgetts) [inline cmd/manager/main.go:81]: Can we revert this back to INFO? I think DEBUG for the default logging level is too verbose

File: docs/docs/guides/oauth2.md
Comment (kylehodgetts) [inline docs/docs/guides/oauth2.md:173]:
```suggestion
* The commands assume you are using Linux.
```

File: internal/controllers/envoyfleet_config_template.go
Comment (kylehodgetts) [inline internal/controllers/envoyfleet_config_template.go:48]: harcoded secrets in a template?

File: internal/controllers/envoyfleet_config_template.go
Comment (kylehodgetts) [inline internal/controllers/envoyfleet_config_template.go:52]: harcoded secrets in a template?

File: internal/envoy/auth/ext_authz_filter.go
Comment (kylehodgetts) [inline internal/envoy/auth/ext_authz_filter.go:39]:
```suggestion
func NewHTTPExternalAuthorizationFilter(upstreamHostname string, upstreamPort uint32, clusterName string, pathPrefix string, authHeaders []*envoy_config_core_v3.HeaderValue) (*anypb.Any, error) {
```
Perhaps?

File: internal/envoy/auth/ext_authz_filter.go
Comment (kylehodgetts) [inline internal/envoy/auth/ext_authz_filter.go:99]: Commented out code

File: internal/envoy/auth/ext_authz_filter.go
Comment (kylehodgetts) [inline internal/envoy/auth/ext_authz_filter.go:115]: Is this supposed to be commented out?

File: internal/envoy/auth/filters.go
Comment (kylehodgetts) [inline internal/envoy/auth/filters.go:41]: commented out code

File: internal/envoy/auth/filters.go
Comment (kylehodgetts) [inline internal/envoy/auth/filters.go:36]: Given that it's returning just a pointer pb duration, could TimeoutDefault be an exported const?

File: internal/envoy/auth/oauth2_filter.go
Comment (kylehodgetts) [inline internal/envoy/auth/oauth2_filter.go:58]: commented out code

File: internal/envoy/auth/oauth2_filter.go
Comment (kylehodgetts) [inline internal/envoy/auth/oauth2_filter.go:78]: commented out code

File: internal/envoy/auth/oauth2_filter.go
Comment (kylehodgetts) [inline internal/envoy/auth/oauth2_filter.go:89]: I think this else block can be removed

File: internal/envoy/auth/oauth2_filter.go
Comment (kylehodgetts) [inline internal/envoy/auth/oauth2_filter.go:119]: commented out code

File: internal/envoy/auth/oauth2_filter.go
Comment (kylehodgetts) [inline internal/envoy/auth/oauth2_filter.go:140]: commented out code

File: internal/envoy/auth/oauth2_parser.go
Comment (kylehodgetts) [inline internal/envoy/auth/oauth2_parser.go:48]: Im not sure if I like this flow. The function is `Parse...` but it's doing more than just that, instead of returning a result of the parsing, it's adding something to the HTTPConnectionManagerBuilder.

Perhaps this function should return the filter and then where is `ParseOAuth2Options` called, the result is then adding to ParseOAuth2Options there?

File: internal/envoy/config/config.go
Comment (kylehodgetts) [inline internal/envoy/config/config.go:175]: What exactly does AutoSni do?

File: internal/envoy/config/hcm.go
Comment (kylehodgetts) [inline internal/envoy/config/hcm.go:226]: commented out code

File: internal/envoy/config/hcm.go
Comment (kylehodgetts) [inline internal/envoy/config/hcm.go:231]: commented out code

File: internal/envoy/manager/envoy_callbacks.go
Comment (kylehodgetts) [inline internal/envoy/manager/envoy_callbacks.go:78]: Do we really want to be flooding the logs with these events if we are not doing anything with them? 
Would it be better to just `return`?

**Review by kylehodgetts — CHANGES_REQUESTED** [review]

File: internal/controllers/envoyfleet_config_template.go
Comment (mbana) [reply internal/controllers/envoyfleet_config_template.go:48]: I've changed this to `<stub_token_secret>` because it needs to be specified so that the user can late adjust it. I really don't think it's a good idea to have them specify/paste:

```yaml
  secrets:
  - name: token
    generic_secret:
      secret:
        inline_string: "<some_token>"
```

Having them change one line is better.

What do you think @kylehodgetts?

File: internal/controllers/envoyfleet_config_template.go
Comment (mbana) [reply internal/controllers/envoyfleet_config_template.go:52]: This is intentional as the user doesn't really have to change this.

If you look at the `GenerateHMAC` function in `internal/envoy/auth/oauth2_filter.go` you'll see it's just 32 random bytes. It's not necessary to run `head -c 32 /dev/urandom | base64` to get this filter to work.

I'll leave this as-is.

What do you think @kylehodgetts.

File: internal/envoy/auth/ext_authz_filter.go
Comment (mbana) [reply internal/envoy/auth/ext_authz_filter.go:39]: Resolved.

I've renamed it to `NewFilterHTTPExternalAuthorization`.

I prefer naming it because if you use an IDE and do completion on `NewFilter<trigger_completion_occurs_here>`, you'll get:

* `NewFilterHTTPExternalAuthorization`
* `NewFilterHTTPOAuth2`.

Thanks.

File: internal/envoy/auth/ext_authz_filter.go
Comment (mbana) [reply internal/envoy/auth/ext_authz_filter.go:99]: Resolved. Remove commented out code. Thanks.

File: internal/envoy/auth/ext_authz_filter.go
Comment (mbana) [reply internal/envoy/auth/ext_authz_filter.go:115]: Resolved. Remove commented out code. Thanks.

File: internal/envoy/auth/filters.go
Comment (mbana) [reply internal/envoy/auth/filters.go:36]: I'll change it to an internal function. The function is pure (it returns the same value each time) so I don't see the point in changing it to a `const`. Is that OK, @kylehodgetts.

File: internal/envoy/auth/oauth2_filter.go
Comment (mbana) [reply internal/envoy/auth/oauth2_filter.go:78]: Removed. Resolved. Thanks.

File: internal/envoy/auth/oauth2_filter.go
Comment (mbana) [reply internal/envoy/auth/oauth2_filter.go:89]: I'd rather keep this log line here, if that OK's with you, @kylehodgetts.

File: internal/envoy/auth/oauth2_filter.go
Comment (mbana) [reply internal/envoy/auth/oauth2_filter.go:119]: Resolved. Removed. Thanks.

File: internal/envoy/auth/oauth2_filter.go
Comment (mbana) [reply internal/envoy/auth/oauth2_filter.go:140]: Resolved. Removed. Thanks.

File: internal/controllers/envoyfleet_config_template.go
Comment (jasmingacic) [reply internal/controllers/envoyfleet_config_template.go:48]: make it so. Replace the token with a place holder

File: internal/controllers/envoyfleet_config_template.go
Comment (jasmingacic) [reply internal/controllers/envoyfleet_config_template.go:52]: Leave it as it is

File: internal/envoy/auth/filters.go
Comment (jasmingacic) [reply internal/envoy/auth/filters.go:41]: Can't we have this documented elsewhere?

File: internal/envoy/auth/filters.go
Comment (jasmingacic) [reply internal/envoy/auth/filters.go:36]: Leave it as it is. If you want add a comment on the function to indicate what you just said

File: internal/envoy/auth/oauth2_filter.go
Comment (jasmingacic) [reply internal/envoy/auth/oauth2_filter.go:89]: Can't you put this into the logger and label it as debug so it is easy to turn on/off

File: internal/envoy/auth/oauth2_parser.go
Comment (mbana) [reply internal/envoy/auth/oauth2_parser.go:48]: I'd rather keep it as it is, if that's OK with you. The reason I'm doing it there is because subtle errors can occur if the filter isn't added properly and in the right order, the documentation of:

```
(h *HCMBuilder) AddFilter(newFilter *hcm.HttpFilter) error
```

Reads:

```
// AddFilter appends f to the list of filters for this HTTPConnectionManager.
// f may be nil, in which case it is ignored and an error will be returned.
// Note that Router filters (filters with TypeUrl `type.googleapis.com/envoy.extensions.filters.http.router.v3.Router`)
// are specially treated. There may only be one of these filters, and it must be the last.
// AddFilter will ensure that the router filter, if present, is last, and will panic
// if a second Router is added when one is already present.
```

File: internal/envoy/auth/oauth2_filter.go
Comment (kylehodgetts) [reply internal/envoy/auth/oauth2_filter.go:89]: If it must stay then it should be part of a logger with the proper level. Error level if something went wrong

File: internal/envoy/config/config.go
Comment (mbana) [reply internal/envoy/config/config.go:175]: TLS is required to reach the `token_endpoint` which is generally on HTTPS. So this `cluster` requires certain additional configuration parameters such as `AutoSni` or specifying the `sni` directly, which requires more code so I opted to just use the `AutoSni` feature. The value of `sni` would need to be calculated and if you read the the function documentation of:

```
// AddClusterWithTLS - AddCluster with SNI or rather `AutoSni` enabled`.
// Example SNI : "kubeshop-kusk-gateway-oauth2.eu.auth0.com" -> "eu.auth0.com"
(e *EnvoyConfiguration) AddClusterWithTLS(clusterName, upstreamServiceHost string, upstreamServicePort uint32) error
```

You have to deduce `"eu.auth0.com"` from `"kubeshop-kusk-gateway-oauth2.eu.auth0.com"`.

I've documented what `AutoSni` does, and a safety measure added a call to `cluster.ValidateAll()` in the `AddClusterWithTLS` function to the `cluster` that has `AutoSni` set to true.

File: internal/envoy/config/hcm.go
Comment (mbana) [reply internal/envoy/config/hcm.go:231]: Resolved. Thanks.

File: internal/envoy/manager/envoy_callbacks.go
Comment (mbana) [reply internal/envoy/manager/envoy_callbacks.go:78]: I think it's quite useful. You could always filter out the logging by using `grep -v OnStreamResponse` etc, but it's not possible to put the logging back in without writing code.

Make the call, do you want me to remove it or shall I leave it there (my preference), @kylehodgetts

Comment (kylehodgetts) [reply internal/envoy/manager/envoy_callbacks.go:78]: We need to think about this from a user perspective. If they are checking the manager logs and they see this, does it help them?

Comment (mbana) [reply internal/envoy/manager/envoy_callbacks.go:78]: I've turned them into debug level log statements then. Please mark as resolved, if you're OK with that.

File: internal/envoy/config/hcm.go
Comment (mbana) [reply internal/envoy/config/hcm.go]: Resolved. Thanks.

File: internal/envoy/auth/ext_authz_filter.go
Comment (kylehodgetts) [reply internal/envoy/auth/ext_authz_filter.go]: Good point 👍

File: internal/controllers/envoyfleet_config_template.go
Comment (mbana) [reply internal/controllers/envoyfleet_config_template.go]: It's been replaced with `<stub_token_secret>` in `internal/controllers/envoyfleet_config_template.go` and in the `docs/docs/guides/oauth2.md` guide.

File: internal/envoy/config/config.go
Comment (jasmingacic) [reply internal/envoy/config/config.go:179]: that's plenty

File: internal/envoy/auth/filters.go
Comment (mbana) [reply internal/envoy/auth/filters.go:36]: Renamed to `timeoutDefault`.

File: internal/envoy/auth/filters.go
Comment (mbana) [reply internal/envoy/auth/filters.go:42]: Where?

Shall I just remove the comment and the code block?

File: internal/envoy/auth/oauth2_filter.go
Comment (mbana) [reply internal/envoy/auth/oauth2_filter.go:73]: It's a a debug level statement now.

It can't be an issue if two clusters exists, it was just a log line I added to help be figure out what was going on. Case in point, if you have two `paths`, like so:

```yaml
    paths:
      "/":
        get:
          description: Returns GET data.
          operationId: "/get"
          responses: {}
      "/uuid":
        get:
          description: Returns UUID4.
          operationId: "/uuid"
          responses: {}
```

then two OAuth2 filters are created, so on the second round the cluster already exists.

File: internal/envoy/auth/filters.go
Comment (jasmingacic) [reply internal/envoy/auth/filters.go:42]: Let's close this. It is taking more time than we want to

File: internal/envoy/auth/oauth2_filter.go
Comment (jasmingacic) [reply internal/envoy/auth/oauth2_filter.go:73]: Either way instead of using fmt.Println can't you use logger to print this out. Having debug println in the code doesn't make us look good.

Comment (mbana) [reply internal/envoy/auth/oauth2_filter.go:73]: `args.Logger.V(1).Info` is a debug level log statement. Marking as resolved.

**Review by jasmingacic — APPROVED** [review]
Once the CI passes please have this merged

**Comment by mbana** [comment]
> Once the CI passes please have this merged

@povilasv has reviewed it and I've addressed the points.

I will merge once the CI passes.

**Review by jasmingacic — APPROVED** [review]

**Review by kylehodgetts — APPROVED** [review]

**Pull request merged**
